packages/@atjson/document/src/index.ts

```diff
@@ -375,6 +375,63 @@ export default class Document {
     return matches;
   }
 
+  canonical(): Document {
+    let canonicalDoc = this.clone();
+    canonicalDoc.where({type: '-atjson-parse-token'}).update(a => {
+      canonicalDoc.deleteText(a.start, a.end);
+    });
+    canonicalDoc.where({type: '-atjson-parse-token'}).remove();
+    canonicalDoc.annotations = canonicalDoc.annotations.sort(this.sortByPositionAndTypeFunction);
+    return canonicalDoc;
+  }
+
+  sortByPositionAndTypeFunction(a: Annotation<any>, b: Annotation<any>): number {
+    let startDelta = a.start - b.start;
+    let endDelta = a.end - b.end;
+    if (startDelta === 0) {
+      if (endDelta === 0) {
+        return a.type.localeCompare(b.type);
+      }
+      else {
+        return endDelta;
+      }
+    } else {
+      return startDelta;
+    }
+  }
+
+  isEqual(docToCompare: Document): boolean {
+    let canonicalLeftHandSideDoc = this.canonical();
+    let canonicalRightHandSideDoc = docToCompare.canonical();
+
+    let isContentEqual = canonicalLeftHandSideDoc.content === canonicalRightHandSideDoc.content;
+    if (!isContentEqual) {
+      return false;
+    }
+    let isAnnotationLengthEqual = canonicalLeftHandSideDoc.annotations.length === canonicalRightHandSideDoc.annotations.length;
+    if (!isAnnotationLengthEqual) {
+      return false;
+    }
+
+    return canonicalLeftHandSideDoc.annotations.every((lhsAnnotation, index) => {
```

Comment:
This comparison function doesn't actually deeply recurse through the attributes (instead, it does a shallow comparison). We need to have a recursive function to compare attributes for this.

Reply:
It may be simpler to add a comparison method on the `Annotation` class that will do this comparison.

Reply:
Can you explain this a bit more -- 

When comparing the attributes of annotations I recursively call isEqual on line 426. I also test the case with nested documents down below. 

What am I missing here :)